src/native/fs/fs_napi.cpp

```diff
@@ -1744,6 +1747,43 @@ struct FileFlock : public FSWrapWorker<FileWrap>
     }
 };
 
+struct FileFcntlLock : public FSWrapWorker<FileWrap>
+{
+    struct flock fl;
+    FileFcntlLock(const Napi::CallbackInfo& info)
+        : FSWrapWorker<FileWrap>(info)
+        , fl()
+    {
+        // lock entire file
+        fl.l_whence = SEEK_SET;
+        fl.l_start = 0;
+        fl.l_len = 0;
+        fl.l_pid = 0;
+        fl.l_type = F_RDLCK;
+
+        if (info.Length() > 1 && !info[1].IsUndefined()) {
+            auto mode = info[1].As<Napi::String>().Utf8Value();
+            if (mode == "EXCLUSIVE") {
+                fl.l_type = F_WRLCK;
+            } else if (mode == "UNLOCK") {
+                fl.l_type = F_UNLCK;
+            } else if (mode == "SHARED") {
+                fl.l_type = F_RDLCK;
+            } else {
+                SetError("invalid lock type");
+            }
+        }
+
+        Begin(XSTR() << "FileFcntlLock" << DVAL(_wrap->_path));
+    }
+    virtual void Work()
+    {
+        int fd = _wrap->_fd;
+        CHECK_WRAP_FD(fd);
+        SYSCALL_OR_RETURN(fcntl(fd, F_OFD_SETLKW, &fl));
```

Comment:
Lets switch to F_SETLKW and add a code comment that explains that the scope of locking we can expect to use is per process and per file, but not per fd, which means that closing any other fd of that file from the same process will unlock it. we will have to workaround this restriction to avoid having two locking flows of the same file in the same process running at the same time.

Reply:
ok, so based on testing this does seem to work with F_OFD_SETLKW.

Reply:
@tangledbytes it would be nice if you could add the test code to the repo or maybe as a comment on this PR, for future reference.